### PR TITLE
Rails Recipe App

### PR DESCRIPTION
**Pull request by albertkantwi**
Within this pull request, we have incorporated the following modifications;

- Integrated Devise for user authentication.
- Implemented authorization using CanCanCan for controlling user access.
- Added models for User, Food, Inventory, InventoryFood, Recipe, and RecipeFood.
- Set up associations between models to represent relationships.
- Implemented controllers for managing foods, recipes, and inventories.
- Created views for displaying lists of foods, recipes, and inventories.
- Designed forms for adding new foods and recipes.
- Developed unit tests for models to ensure proper validations.
- Implemented request specs for controllers to verify route configurations.
- Styled views for a visually appealing user interface.
- Ensured mobile responsiveness for a seamless experience on various devices.
- Fixed linters' errors.

Kindly review for us.

### CODE REVIEW

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<p>Welcome <%= @email %>!</p>
+
+<p>You can confirm your account email through the link below:</p>
+
+<p><%= link_to 'Confirm my account', confirmation_url(@resource, confirmation_token: @token) %></p>
```
Comment (mukhammadrizooff) [inline app/views/devise/mailer/confirmation_instructions.html.erb:5]: - [ ] [REQUIRED] Good job so far! I appreciate what you have done but the email registered in the `"Login"` and `"Signup"` parts is not confirmed, if the registered email is not confirmed, it is not possible to go from the `"Login"` page to other pages. This does not allow you to see all the properties of this project, so please fix this error 🙏🏼 

image

**Review by mukhammadrizooff — CHANGES_REQUESTED** [review]
Hi @TEAM, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights
- TEAM work ✔️ 
- No linter errors ✔️ 
- Correct GitFlow ✔️ 
- Good implementations ✔️
- Descriptive `README` file ✔️
- All Spec file's tests passed ✔️ 
image


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/recipes/index.html.erb
Comment (aamirkhan2478) [inline app/views/recipes/index.html.erb:1]: - [ ] Make sure your Recipes page look like a wireframe but for now your Recipes page page is not look like wireframe check below :point_down: to check comparison between your design and wireframe

**Your Design**
screencapture-localhost-3000-recipes-2023-12-01-00_19_14

**Wireframe**
wireframe

File: app/views/recipes/public_recipies.html.erb
Comment (aamirkhan2478) [inline app/views/recipes/public_recipies.html.erb:1]: - [ ] Make sure your Public Recipes page looks like a wireframe but for now, your Public Recipes page page does not look like a wireframe check below 👇 to check the comparison between your design and wireframe

**Your Design**
screencapture-localhost-3000-public-recipies-2023-12-01-00_21_05

**Wireframe**
Wireframe

File: app/views/recipes/show.html.erb
Comment (aamirkhan2478) [inline app/views/recipes/show.html.erb:1]: - [ ] Make sure your Recipes Details page looks like a wireframe but for now, your Recipes Details page page does not look like a wireframe check below 👇 to check the comparison between your design and wireframe

**Your Design**
screencapture-localhost-3000-recipes-1-2023-12-01-00_25_41

**Wireframe**
wireframe

File: app/views/shopping_list/index.html.erb
Comment (aamirkhan2478) [inline app/views/shopping_list/index.html.erb:1]: - [ ] Make sure your Public Recipes page looks like a wireframe but for now, your Public Recipes page page does not look like a wireframe check below 👇 to check the comparison between your design and wireframe

**Your Design**
screencapture-localhost-3000-shopping-list-2023-12-01-00_28_10

**Wireframe**
wireframe

**Review by aamirkhan2478 — CHANGES_REQUESTED** [review]
Hi @albertkantwi & @pushkar1995, 



Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

 ## Highlights
✔️ Good Effort
✔️ Your design is good.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: Gemfile [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  gem 'capybara'
+  gem 'database_cleaner'
+  gem 'launchy'
+  gem 'selenium-webdriver'
+  gem 'webdrivers'
+end
+gem 'rails-controller-testing'
+gem 'rubocop', '>= 1.0', '< 2.0'
+
+gem 'bootstrap', '~> 4.0'
```
[the remark this answers is not in the document]
Comment (rloterh) [reply Gemfile]: Working properly! ✔️

File: app/views/recipes/index.html.erb
Comment (rloterh) [reply app/views/recipes/index.html.erb:1]: Resolved, but can be improved further. Let's ensure to add adequate spacing around the elements in this page, as seen in the design [template](https://raw.githubusercontent.com/microverseinc/curriculum-rails/main/recipe-app/images/recipe_index.png?token=GHSAT0AAAAAACFSKGIJHITG747ALQK4U4GGZLK3KPQ)

File: app/views/recipes/public_recipies.html.erb
Comment (rloterh) [reply app/views/recipes/public_recipies.html.erb:1]: Properly resolved! ✔️

File: app/views/recipes/show.html.erb
Comment (rloterh) [reply app/views/recipes/show.html.erb:1]: Properly resolved! ✔️

File: app/views/shopping_list/index.html.erb
Comment (rloterh) [reply app/views/shopping_list/index.html.erb:1]: Properly resolved! ✔️

**Review by rloterh — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi Albert & Pushka,  

**RoR group project** is complete! 🏁 Good job! 👏 👏  Congratulations!  🎉🌟 

## Highlights 🎯 
🎲 Authentication with Devise gem properly implemented!
🎲 App schema closely follows given ERD!
🎲 Only particular owners can delete contents added by himself/herself!
🎲 All users can view public recipes!

## Optional suggestions 🧰 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

**You can also consider:**
Let's make the ReadMe stand out even better by including instructions under the _Getting Started_ section, on how to run the app, for instance, using the command `rails server`.


Keep being awesome! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**